### ISSUE TITLE
Templating as part of broader generative experience

### ISSUE THREAD

**Issue by KathleenDollard** [dotnet/templating#3954]
# .NET Templating as part of broader generative experience

tl;dr; _We have the mechanisms to an experience where users make preliminary decisions on template run, and can then very safely change these experiences as they need. The ways we would alter architectures is the only hard part of the effort, but might align very well with other simplification efforts. A specific use case to consider is authentication._ 

This proposal draws from suggestions and work from Vlada Shubina, David Karlas, Bekir Ozturk, Sayed Hashimi, Phil Henning, Tim Heuer, Brady Gaster and Glenn Condron and others. I think most folks think that this would be a better experience. This paper is heavily opinionated based on my experience with generative practices. 

We have an effective mechanism for most templates today. This is not a replacement for that. This is instead a way to offer a very good user experience for very common templates that we create. Today’s mechanism is based on “runnable templates.” Any changes to templates should retain runnable templates. User actions: 

* Run a template
* Do work
* Make change 
* Do work
* Make change

The next generation of .NET templating can: 

* Retain the commitment to runnable templates
* Be re-entrant and support natural/code based metadata and logical flow based templates (such as T4 or Roslyn source generators) when re-entrant
* Work with, encourage, possibly require special architectures  to isolate code managed by the re-entrant template
* Analyzers and code fixers to support the re-entrant template, where appropriate

This paper describes all those weird words, and then gives an example. 

The approach described here relies on the availability of semantic code analysis (such as the ability to find a reference), and thus is currently applicable to the Roslyn languages – C# and VB.

## What is a “runnable template”?

The concept of “runnable templates” is that the template itself is valid code that can be run in place. This is meaningful only in project templates, because the result of an item template is not runnable by itself. It means that the directory of files the template is built from is a runnable project. This has tremendous value in the ease of creating templates. It also implicitly excludes any (known) logical flow templating mechanism. 

The proposal here is to retain the current templating mechanism. Instead of altering this mechanism, add a follow-on re-entrant mechanism. 

When the user creates the project, template execution would be the current mechanism, and then run a build. The build would run source generators that would also run on future builds. We already do a restore so there will be very little performance hit. Since the source generators also run on future builds, the user merely needs to change their code and generation will run. 

Declarations, anything currently a parameter would become part of the user codes. Examples of what this might look like would be specific class (“new AuthNone(” vs “new AuthAzureMagic(”), method, property and enum usage and/or declaration. 

## What does re-entrant mean in this context?

A re-entrant template is one that you can run again with different metadata and it will alter your code, removing code from the previous run of the template and adding code for the new run. It alters your code in place. See the section Special Architectures. 

It’s also a terrible name. It just happens, the user doesn’t need to deliberately rerun, and it is only a template in the broad sense because it’s a source generator. Its purpose is to apply changes. 

## What is a “logical flow based template”?

Another terrible name I made for the concept of not-runnable and capable of complex analysis. Maybe in this context it should just be called a generator, but I tried to lay this out to encourage alternate thinking – maybe everything is not a generator. Anyway…

Runnable templates have severe limitations. As they exist today, they can’t do loops. Conditionals are via #ifdef so must include blocks of code and are awkward. You can’t build a line of code based on logic so things are simple as the name of a method is impossible. 

A logical flow based template is a template that allows logic, and is not runnable. The common historic example is T4, but there is no reason it can’t be an application. In fact the most important approach to T4 is executable T4 where T4 creates source code that when compiled creates source code. Roslyn source generators are another example of a logical flow based template.

For this project, they aren’t actually a template, they are source generators. Which means, we need no special mechanism to run them. 

_(NOTE: Years ago I mostly wrote a templating language that used additional restrictions on C# to create a templating language. It is not impossible, there are just sufficient limitations a runnable flow based language has to make that it is not practical, and at best requires arcane knowledge. And, at least my approach at the time, required external metadata and successful generative processes today rely on code as metadata. That failure led me to believe that runnable and logical flow templates are exclusive concepts in C# and VB.)_

### Should this effort use T4?

T4 solves a very small part of the problem addressed here. Once you have an architecture (see below) that supports generation, a runnable template to lay out the basic structure and get things ready, a mechanism to rerun the template, and a mechanism to evaluate source code and build the model for generation you’ve done the hard work. Whether the step of generating text is via T4, or interpolated strings, or a set of methods that limit the ways you can mess up the syntax is a small part of the problem. Using T4 will require being able to evaluate a T4 template and that is messy code. The question of T4 both isn’t very important, and we can probably do better for C#/VB code with 2 dozen specialized methods in a library that we can share, and interpolated strings for non C#/VB stuff. 

## What is a “re-entrant template”

It is a generative process that uses the existing code and possibly configuration files to rewrite portions of the code. It requires the ability to find previously generated code to re-write it (destructive generation) and to understand the non-generated code to determine how to re-write.  

This paper is full of bad names. It means it can be run more than once. In this case, there are two templates: A simplified template using the mechanism we know and love today and a reentrant part. In this paper, it is a just a generator that just runs on build.

## What is “natural/code based metadata”

Metadata is the data that drives generation. It is how the user indicates details of the output. Natural metadata is metadata that the user would already create. Examples are databases and source code. Databases are external, in that they are not part of the generated code. User source code is internal metadata in that it is part of the code that is being generated. For example, Auth type and details could be declared in user code. Another example is (a reasonable number of ) parameters passed when `dotnet new` runs

This is in contrast to contrived metadata that exists only for the sake of the generator. Contrived metadata may be necessary when a project is first created. Contrived metadata adds complexity and ideally we would avoid it by requiring no metadata that the user hasn’t entered for another purpose.  

Code based metadata is very important. Roslyn gave us this, although it isn’t easy. Part of the work of this project can be to establish easier ways to understand code from Roslyn (where possible via the code neutral semantic layer, allowing generator guts to be cross language).

## Special architectures

There are many things we can do with templates, but re-entrant templates should only be done with an expectation of altering architectures because safety of reentrancy.

**This is the only hard part of this proposal. Everything else is just work. And these architectures are not radical and will benefit users.**

### Isolation

This section makes the case for the criticality of isolation. If you don’t need to be convinced, skip it.

A principle of code generation is that user code is sacred. It should never be compromised. This is an absolute statement and is approximately impossible to achieve without isolating user code from generated code. If they are together, the user can and will reorder unimportant code, interleave important code of their own and alter the wrapping to suit their tastes. A very sophisticated generator can sort this out (except perhaps the wrapping) _if it knows what the original generated code was_. To know what the original code was requires either maintaining a copy of the code immediately after generation, or knowing all of the metadata (including potentially environmental metadata from another machine if the user receives the code by cloning a repo) and the template for generation. Oh, and you need to have the initial generator as well. Thus, it is very difficult, but possible to do mixed code in Version One of your template. By Version Six, you must also have all previous templates and an indicator of which template was used. Over time mixed output becomes approximately impossible to maintain. Especially when there is an alternative that is also better for the user. 

I have heard _we will recognize whether the user has changed the code and only update the code if they have not_. There are two problems with this statement. The first is that a minor code change like reordering lines should not block future actions, especially since there is no way for the user to know at the time they make the change that they have just hurt their future selves. The second is that it requires you know what template and metadata was used in the initial generation to recognize the code – see earlier discussion on template versioning (or saving the code to a safe location, or saving a hash of the code, all of which is also vulnerable to use deletion). 

I have also heard _we will add a comment warning the user not to change the code_. That has a small problem and a big problem. The small problem is that users do not read comments. The big problem is that unless you put the comment between every line of code the user will never see the comment. The predominate ways we navigate source code do not start at the top.

Isolating generated code is better for the user. Partial classes were introduced for this purpose in .NET in around 2005, and partial classes received an important update alongside Roslyn source generators. One of those updates was that users can’t edit generated files. There is no way to provide this security without separate files.

### Adjusting architectures

Adjusting architectures to embrace re-entrant generation is good for users. We are already doing this very heavily with Roslyn source generators and other changes making ASP.NET architectures much simpler. It is a good time to look at it this in relation to templates. 

A good example is authentication. If we can’t solve auth, then, we failed. Auth has the highest value to users, because asking to do auth when you create your project is silly. They need to do this later once they get things figured out.

### An Example: Auth

Auth has a set of inputs, including what kind of auth and depending on a set of additional information. This information can be placed in a declarative way as part of appropriate code. In the default case this is the only code the user needs to bother with and this is the state after the initial template generation (before the re-entrant generation step). The re-entrant generation runs as source generators in the compile. The user starts their project. 

(This part is vague as it needs ASP.NET input. I would like to sit down with one of the complex templates and walk through what it would look like with this approach and identify challenges.)

There are several kinds of changes the user will make in the lifetime of their project. There are probably files that the user will just change, such as customizing a login page. These may not need to be part of generation, except possibly to offer to remove the file as dead code after an auth change. 

There will also be methods called that the user may want to customize. A default implementation can be generated if, and only if, the user has not provided an implementation (this is how the new partial methods work).

At some point, the user wants to change their authentication. They can either just change it. Any additional information will be obvious and they fix it. Most or all code and information no longer needed just drops or errors. 

One way to implement this would be an AuthInfo base class. Each mechanism for auth would be a derived class with a constructor that with required and optional parameters for all of the information needed for that style of auth. Thus the user would naturally supply the correct information and allow customizations of auth behavior by overriding methods in the specific AuthInfo class. 

With a base/derived class implementation of auth styles, the user changes auth styles by changing the class they are instantiating, and changing therefore changing the constructor that is called. If required information is missing or a parameter is passed that is not valid, the user gets an immediate error and can easily understand why this occurred. 

_Note: This will be much safer if the user elects to used named parameters so Boolean and ints aren’t misunderstood. This is a great case for the future C# feature of properties that are required on initialization, and in that case using initializers instead of constructors. However, that feature doesn’t yet exist so I suggest constructors here._

Analyzers and code fixes can be used to help find dead code that results from changes. This will occur when the user writes a method that is treated as a partial that would normally replace generated code. If the user changes auth so that the method is no longer needed, we may be able to recognize this and remove the unused code. The uncertainty here is only whether the old library would still be present to perform the analysis. If the auth mechanisms are in separate libraries, and the user removes the library before compiling, the dead code may not be flagged. Hmm. Issues to consider.

## And we still need analyzers?

See the section “An example” for an example.

## Isn’t this overly complicated?

Not for users. It optimizes for a great, simple user experience. 

Simple templates are handled well today. This is designed for a known set of problems in Microsoft templates where Roslyn source are in use and the current massively complex templates and lack of ability for users to change their mind is a known problem. Roslyn source generators are a good solution to the problem of reentrancy. The mechanism for generators themselves may be slightly arcane, but is simple. Evaluating code to figure out the code to generate can sometimes be complex. 

## What this does not cover

There are a few classes of problems that this does not handle:

•	Generation based on database metadata: Databases are guaranteed to change, so solutions here need to be re-entrant. I do not think Roslyn source generators have a way to evaluate a database, and it would destroy build performance. But, EF is solving this.
•	Logic during initial template execution.
•	Loads of other stuff I have not thought of.

**Comment by marcpopMSFT** [comment]
This issue is being closed as won't fix. It has been open for an extended period without maintainer activity or community demand, and is unlikely to be addressed.

The dotnet/templating repository has been consolidated into [dotnet/sdk](https://github.com/dotnet/sdk). If this issue is still important to you, please open a new issue in [dotnet/sdk](https://github.com/dotnet/sdk/issues/new) with updated context so the team can re-evaluate.

<details>
<summary>Triage details</summary>

**Summary:** Design proposal for re-entrant templating with source generators as part of a broader generative experience

**Reasoning:** This is a high-level design proposal from 2021 about future templating experiences using source generators and re-entrant templates. It has zero upvotes, zero comments, no activity since September 2021 (nearly 5 years old), and the repo is being consolidated into dotnet/sdk. The proposal was never acted upon and is stale.

</details>
